### PR TITLE
feat(cli): bind model classes during boot

### PR DESCRIPTION
**Pull request by raymondfeng**
## Checklist

👉 [Read and sign the CLA (Contributor License Agreement)](https://cla.strongloop.com/agreements/strongloop/loopback-next) 👈

- [x] `npm test` passes on your machine
- [x] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- [x] API Documentation in code was updated
- [ ] Documentation in [/docs/site](../tree/master/docs/site) was updated
- [ ] Affected artifact templates in `packages/cli` were updated
- [ ] Affected example projects in `examples/*` were updated

👉 [Check out how to submit a PR](https://loopback.io/doc/en/lb4/submitting_a_pr.html) 👈

### CODE REVIEW

File: packages/boot/src/booters/model.booter.ts [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    }
+  }
+}
+
+/**
+ * Default ArtifactOptions for DataSourceBooter.
+ */
+export const ModelDefaults: ArtifactOptions = {
+  dirs: ['models'],
+  extensions: ['.model.js'],
```
Comment (achrinza) [inline packages/boot/src/booters/model.booter.ts:69]: Ditto

```suggestion
  extensions: ['.model.ts'],
```

File: packages/boot/src/__tests__/integration/model.booter.integration.ts [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    );
+
+    await sandbox.copyFile(
+      resolve(__dirname, '../fixtures/multiple-models.model.js'),
+      'models/multiple-models.model.js',
+    );
+
+    const MyApp = require(resolve(sandbox.path, 'application.js')).BooterApp;
+    app = new MyApp();
+  }
```
Comment (achrinza) [inline packages/boot/src/__tests__/integration/model.booter.integration.ts:52]: Shouldn't the fixtures' extension be `.ts`?

```suggestion
  async function getApp() {
    await sandbox.copyFile(resolve(__dirname, '../fixtures/application.ts'));
    await sandbox.copyFile(
      resolve(__dirname, '../fixtures/no-entity.model.ts'),
      'models/no-entity.model.ts',
    );
    await sandbox.copyFile(
      resolve(__dirname, '../fixtures/product.model.ts'),
      'models/product.model.ts',
    );

    await sandbox.copyFile(
      resolve(__dirname, '../fixtures/multiple-models.model.ts'),
      'models/multiple-models.model.ts',
    );

    const MyApp = require(resolve(sandbox.path, 'application.ts')).BooterApp;
    app = new MyApp();
  }
```

File: packages/repository/src/model.ts [hunk 3]
```diff
@@ -214,7 +214,7 @@ function asObject(value: any, options?: Options): any {
 /**
  * Base class for models
  */
-export abstract class Model {
```
Comment (achrinza) [inline packages/repository/src/model.ts:217]: Any reason it's no longer abstract? IIUC, the `Model`, `Entity` and `ValueObject` classes should only be extended from and not be used directly.

File: packages/boot/src/__tests__/integration/model.booter.integration.ts (hunk 2 above)
Comment (raymondfeng) [reply packages/boot/src/__tests__/integration/model.booter.integration.ts:52]: @achrinza We always only scan the transpiled `.js` code as the booters need to load them.

File: packages/repository/src/model.ts (hunk 3 above)
Comment (raymondfeng) [reply packages/repository/src/model.ts:217]: TypeScript does not allow using `Model` as a value (class) if it's abstract. I can revert these changes and only test if the class is decorated with `@model`.

File: packages/boot/src/__tests__/integration/model.booter.integration.ts (hunk 2 above)
Comment (achrinza) [reply packages/boot/src/__tests__/integration/model.booter.integration.ts:52]: @raymondfeng Ah, gotcha. Slipped my mind that it'll already be transpiled, my bad 🙇

File: packages/boot/src/booters/model.booter.ts [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+  /**
+   * Uses super method to get a list of Artifact classes. Boot each file by
+   * creating a DataSourceConstructor and binding it to the application class.
+   */
+  async load() {
+    await super.load();
+
+    for (const cls of this.classes) {
+      if (!isModelClass(cls)) continue;
```
Comment (bajtos) [inline packages/boot/src/booters/model.booter.ts:49]: Please print a `debug` log when skipping a class that's not recognized as a model.

**Comment by pookdeveloper** [comment]
I need it in my project that I use, please review it as soon as you can thanks! :)

**Comment by pookdeveloper** [comment]
I need it in my project that I use, please review it as soon as you can thanks!

File: packages/boot/src/booters/model.booter.ts [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+  /**
+   * Uses super method to get a list of Artifact classes. Boot each file by
+   * creating a DataSourceConstructor and binding it to the application class.
+   */
+  async load() {
+    await super.load();
+
+    for (const cls of this.classes) {
+      if (!isModelClass(cls)) {
+        debug('Skipping class %s - no @model is found', cls.name);
```
Comment (hacksparrow) [inline packages/boot/src/booters/model.booter.ts:50]: Do we need to do this? In future we may want to boot other classes which are not models.

Comment (raymondfeng) [reply packages/boot/src/booters/model.booter.ts:50]: @hacksparrow The model booter is for model classes only.

File: packages/boot/src/booters/model.booter.ts [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+        continue;
+      }
+
+      debug('Bind class: %s', cls.name);
+      // We are binding the model class itself
+      const binding = this.app.bind(`models.${cls.name}`).to(cls).tag('model');
```
Comment (bajtos) [inline packages/boot/src/booters/model.booter.ts:56]: @raymondfeng can you please extract the code building the binding to a standalone function that can be used to bind models manually, in cases where `@loopback/boot` is not involved? For example in extensions.

```ts
const binding = createModelClassBinding(cls);
app.add(binding);
```

Similar helpers we already have: `createComponentApplicationBooterBinding`, `createBindingFromClass`, `createBodyParserBinding`, `createServiceBinding`.

**Pull request merged**